### ISSUE TITLE
Area2D + Colision(Shape|Polygon)2D don't restrict mouse events to polygon

### ISSUE THREAD

**Issue by rosshadden** [godotengine/godot#10424]
**Operating system or device - Godot version:**

Arch Linux - Godot `master`

**Issue description:**

Everything I can find on the internet about how to detect clicks inside a node say to use `Area2D` + (`CollisionShape2D` or `CollisionPolygon2D`).  But in the current version of Godot, click events will fire whether they are within the specified polygon or not.
I believe this worked as expected a couple months ago, but I might be misremembering.

**Steps to reproduce:**

- Make the following tree:

```
Area2D
  CollisionPolygon2D
```

- Draw/define a polygon
- Give `Area2D` the following script:

```gdscript
extends Area2D

func _unhandled_input(event):
	if event is InputEventMouseButton and event.button_index == 1 and event.is_pressed():
		print("click")
```

- Click anywhere
  - Shouldn't fire but does
- Click on the polygon (might want to add a sprite, image, or `Polygon2D` as well to make it visible)
  - Should fire and does

**Link to minimal example project:**

This is very far from minimal, as it's not even in gdscript. But https://github.com/acleverpun/tankerz/blob/feature/areas/src/building.nim#L34-L36

**Comment by eon-s** [comment]
I believe that is expected because `_unhandled_input` is a Node method, you should use `_input_event`, the CollisionObject method to detect inputs (for "pickable" objects).

**Comment by RebelliousX** [comment]
I think you can workaround this by getting the world direct space 2d, and check for intersect_point(mouse_pos). So, this way, you can check if the mouse position when clicking for example, is within collision shape/polygon 2d.

**Comment by rosshadden** [comment]
Okay thank you.

**Comment by rosshadden** [comment]
@RebelliousX I just tried what you said, as I think it would be better suited for what I am trying to do (have one click handler rather than n).  However the array of intersected collision shapes is always empty, even though I am clicking inside collision polygons and using that click vector.

[This](https://github.com/acleverpun/tankerz/blob/dev/intersections/scenes/arena.gd) is a branch where I tried it.  Am I doing anything wrong, or do you perhaps have any working code you could show me with what you are talking about?  Thank you.

**Comment by RebelliousX** [comment]
@rosshadden From the documentation, page 667

> • Array intersect_point ( Vector2 point, int max_results=32, Array exclude=Array(), int
layer_mask=2147483647, int type_mask=15 )
Check whether a point is inside any shape. The shapes the point is inside of are returned in an array containing dictionaries with the following fields:
shape: Shape index within the object the point is in.
metadata: Metadata of the shape the point is in. This metadata is different from Object.get_meta, and is set with
Physics2DServer.shape_set_data.
collider_id: Id of the object the point is in.
collider: Object the point is inside of.
rid: RID of the object the point is in.
Additionally, the method can take an array of objects or :ref:‘RID<class_rid>‘s that are to be excluded from collisions, a bitmask representing the physics layers to check in, and another bitmask for the types of objects to check (see TYPE_MASK_* constants).

So basically, you have to use for loop to go through all items in array and check each like `for item in array:` and then `item["collider"]` to get the collider object. I don't think a simple print() will be able to print the array that contains dictionary for each item, or you might see them as null, but there should be some data.
